tests/engagement/test_churn.py

```diff
@@ -383,3 +384,87 @@ def case(name, expect, desktop, mobile, configured):
 
     for test in test_func:
         test(df)
+
+
+def test_attribution_from_new_profile(effective_version,
+                                      generate_main_summary_data,
+                                      generate_new_profile_data):
+    main_summary = generate_main_summary_data([
+        {'client_id': '1', 'attribution': {'source': 'mozilla.org'}},
+        {'client_id': '3', 'attribution': None},
+        {'client_id': '4', 'attribution': None},
+        {
+            'client_id': '6',
+            'attribution': {'source': 'mozilla.org'},
+            'timestamp': SUBSESSION_START.shift(days=1).timestamp * 10 ** 9,
+        },
+        {'client_id': '7', 'attribution': {'source': 'mozilla.org'}},
+    ])
+
+    def update_attribution(attribution):
+        # only updates the attribution section in the environment
+        env = copy.deepcopy(data.new_profile_sample['environment'])
+        env['settings']['attribution'] = attribution
+        return env
+
+    new_profile = generate_new_profile_data([
+        # new profile without a main summary companion
+        {'client_id': '2', 'environment': update_attribution({'source': 'mozilla.org'})},
+        # recover null attribution
+        {'client_id': '3', 'environment': update_attribution({'source': 'mozilla.org'})},
+        # new-profile ping used to recover attribution, but outside of the
+        # the current retention period
+        {
+            'client_id': '4',
+            'environment': update_attribution({'source': 'mozilla.org'}),
+            'submission': SUBSESSION_START.shift(days=-7).format("YYYYMMDD"),
+        },
+        # ping outside of the current retention period not counted toward totals
```

Comment:
This isn't necessarily being tested here, since `5` is not in the main_summary dataset (it's testing what `2` is testing: a new profile without a main summary companion).